httpload.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+type stats struct {
+	failedRequest  int
+	successRequest int
+
+	reqTimes     []time.Duration
+	totalTime    time.Duration
+	slowest      time.Duration
+	fastest      time.Duration
+	average      time.Duration
+	requestsXSec float64
+}
+
+func (s *stats) reqXn(url string, reqNum int, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for i := 0; i < reqNum; i++ {
+		start := time.Now()
+		r, err := http.Get(url)
+		if err != nil {
+			fmt.Println(err.Error())
+			s.failedRequest++
+			return
+		}
+
+		r.Body.Close()
+		s.successRequest++
+		s.elapsedTime(start)
+	}
+}
+
+func (s *stats) reqXtime(cancelCtx context.Context, url string) {
+	client := http.Client{}
+
+	for {
+		start := time.Now()
+		r, err := http.NewRequestWithContext(cancelCtx, "GET", url, nil)
+		if err != nil {
+			fmt.Println(err.Error())
+			s.failedRequest++
+			return
+		}
+
+		resp, err := client.Do(r)
+		if err != nil {
+			err = errors.Unwrap(err)
+			if err == context.DeadlineExceeded {
+				return
+			}
+
+			fmt.Println(err.Error())
+			s.failedRequest++
+			return
+		}
+
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+		s.successRequest++
+		s.elapsedTime(start)
+	}
+}
+
+func (s *stats) elapsedTime(start time.Time) {
+	elapsed := time.Since(start)
+	s.reqTimes = append(s.reqTimes, elapsed)
```

Comment:
We haven't got the time to speak about the race detector.
To enable it it's very easy, just add -race to `go run` -> `go run -race httpload.go`
The program will print on standard output, in case there are any, the line numbers where data is accessed from multiple goroutines at a time.

If i'm saying that, it's because we have a race condition here.
If we have more than one worker for example, the workers will call `elapsedTime`  at the same time. The problem is at line 79 you modify the `reqTimes` slice.
You can keep the current code but you'll have to protect `s.reqTimes` with a [sync.Mutex](https://golang.org/pkg/sync/#Mutex) for example. 

After fixing it, the race detector should not print nothing. But that doesn't mean that you don't have a race condition in your program, that means that the race detector haven't seen it. The race detector is stupid, in the sense that when enabled, it monitors the accesses to *all* variables and prints when it sees a race condition.